types/react/test/managedAttributes.tsx

```diff
@@ -158,3 +158,67 @@ interface LeaveMeAloneDtslint { foo: string; }
 //         reqNode={<span />}
 //     />
 // ];
+
+// class ComponentWithNoDefaultProps extends React.Component<Props> {}
+
+// function FunctionalComponent(props: Props) { return <>{props.reqNode}</> }
+// FunctionalComponent.defaultProps = defaultProps;
+
+// const functionalComponentTests = [
+//     // $ExpectError
+//     <FunctionalComponent />,
+//     // This is possibly a bug/limitation of TS
+//     // Even if JSX.LibraryManagedAttributes returns the correct type, it doesn't seem to work with non-classes
+//     // This also doesn't work with things typed React.SFC<P> because defaultProps will always be Partial<P>
+//     // $ExpectError
+//     <FunctionalComponent str='' />
+// ];
+
+// const MemoFunctionalComponent = React.memo(FunctionalComponent);
+// const MemoAnnotatedDefaultProps = React.memo(AnnotatedDefaultProps);
+// const LazyMemoFunctionalComponent = React.lazy(async () => ({ default: MemoFunctionalComponent }));
+// const LazyMemoAnnotatedDefaultProps = React.lazy(async () => ({ default: MemoAnnotatedDefaultProps }));
+
+// const memoTests = [
+//     // $ExpectError
+//     <MemoFunctionalComponent />,
+//     // $ExpectError won't work as long as FunctionalComponent doesn't work either
+//     <MemoFunctionalComponent str='abc' />,
+//     // $ExpectError
+//     <MemoAnnotatedDefaultProps />,
+//     <AnnotatedDefaultProps str='abc' />,
+//     // $ExpectError this doesn't work despite JSX.LibraryManagedAttributes returning the correct type
+//     <MemoAnnotatedDefaultProps str='abc' />,
+//     // $ExpectError won't work as long as FunctionalComponent doesn't work either
+//     <LazyMemoFunctionalComponent str='abc' />,
+//     // $ExpectError
+//     <LazyMemoAnnotatedDefaultProps />,
+//     // $ExpectError this doesn't work despite JSX.LibraryManagedAttributes returning the correct type
+//     <LazyMemoAnnotatedDefaultProps str='abc' />
+// ];
+
+// type AnnotatedDefaultPropsLibraryManagedAttributes = JSX.LibraryManagedAttributes<typeof AnnotatedDefaultProps, Props>;
+// // $ExpectType AnnotatedDefaultPropsLibraryManagedAttributes
+// type FunctionalComponentLibraryManagedAttributes = JSX.LibraryManagedAttributes<typeof FunctionalComponent, Props>;
+// // $ExpectType FunctionalComponentLibraryManagedAttributes
+// type MemoFunctionalComponentLibraryManagedAttributes = JSX.LibraryManagedAttributes<typeof MemoFunctionalComponent, Props>;
+// // $ExpectType FunctionalComponentLibraryManagedAttributes
+// type LazyMemoFunctionalComponentLibraryManagedAttributes = JSX.LibraryManagedAttributes<typeof LazyMemoFunctionalComponent, Props>;
```

Comment:
Assuming I'm using `$ExpectType` correctly these all pass. It just doesn't _actually_ work when the functional/exotic components are used in JSX.